### PR TITLE
feat: test case coverage for digest, salt

### PR DESCRIPTION
**Pull request by gjj**
For the digest process:

- Added `illegalCharactersCheck`, to be done before `salt`ing the document
  - Checks for `.`, `[` and `]` to prevent such characters from being salted since they would cause shadowed keys
  - I didn't put the checks inside `salt` due to how `${path}.${key}` is being recursively passed into `deepMap`
- Fixed the supposed non-deterministic salts, changed it to using hardcoded salts
- Added `number` and `boolean` as allowed data types, and `null` is accepted now
  - Allowed data types are: `object, array, string, number, boolean` and values: `null`, everything else is rejected
- Improved test cases expected values

For schema:

- Since `null` is accepted, changed some test cases expected values (should now be expecting schema validation errors)

Also added test case coverage for:

- shadowed keys (arrays and objects)
- null and undefined values in document
- sparse arrays

### CODE REVIEW

File: src/v3/digest/digest.test.ts [hunk 1]
```diff
@@ -1,41 +1,150 @@
...
-          type: TemplateType.EmbeddedRenderer,
-          url: "https://renderer.openattestation.com/"
-        },
-        oaProof: {
-          type: OaProofType.OpenAttestationProofMethod,
-          method: Method.DocumentStore,
-          value: ""
+      const document = cloneDeep(sampleDoc);
+
+      // We shouldn't use const salts = salt(document); here as it's non-deterministic
```
Comment (Nebulis) [inline src/v3/digest/digest.test.ts:33]:
```suggestion
      // We shouldn't use create salts on every test; here as it's non-deterministic
```

File: src/v3/digest/digest.test.ts [hunk 2]
```diff
@@ -1,41 +1,150 @@
...
-      const digest = digestDocument(document, salts, []);
-      expect(digest).toBeTruthy();
+    test("handles null values correctly", () => {
+      const document = {
+        ...cloneDeep(sampleDoc),
+        grades: null
+      };
+      expect(() => {
+        salt(document);
+      }).toThrow("Cannot convert undefined or null to object");
```
Comment (Nebulis) [inline src/v3/digest/digest.test.ts:123]: what throw this error? null is a valid json value, so should be ok :grimacing:

Comment (gjj) [reply src/v3/digest/digest.test.ts:123]:
```
  if (typeof value === "object") {
    return Object.keys(value).flatMap(key => deepMap(value[key], path ? `${path}.${key}` : key));
  }
```

looks like that's because the type of `null` values are `object` so the error is caused by this part of `deepMap`. since null values are accepted, we'll update the test cases for schema too as well

File: src/v3/digest/digest.test.ts (hunk 1 above)
Comment (gjj) [reply src/v3/digest/digest.test.ts]: fixed

File: src/v3/digest/digest.test.ts (hunk 2 above)
Comment (gjj) [reply src/v3/digest/digest.test.ts]: ok since `null` is an accepted value, i modified `deepMap` to accept it

also changed schema tests since error messages for null values should have changed (now expecting schema errors instead)

Comment (gjj) [reply src/v3/digest/digest.test.ts]: fixed

- since null is accepted value, changed some tests for schema
- edited deepMap type check (to handle null properly)
- improved digest test case expected values by using toContainEqual+objectContaining

File: src/v3/wrap.ts [hunk 3]
```diff
@@ -13,16 +13,34 @@ const deepMap = (value: any, path: string): Salt[] => {
   if (Array.isArray(value)) {
     return value.flatMap((v, index) => deepMap(v, `${path}[${index}]`));
   }
-  if (typeof value === "object") {
+  if (typeof value === "object" && value) {
```
Comment (Nebulis) [inline src/v3/wrap.ts:16]: let's add a comment on this one to explain the && value

Comment (gjj) [reply src/v3/wrap.ts:17]: added a comment to explain the reasoning
